File: app/src/main/java/org/secuso/privacyfriendlytodolist/view/calendar/CalendarView.java
package org.secuso.privacyfriendlytodolist.view.calendar;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.GridView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import org.secuso.privacyfriendlytodolist.R;
import java.util.Calendar;
import java.util.Date;


public class CalendarView extends LinearLayout {

    private static final int MAX_DAY_COUNT = 42;
    private static final String DATE_FORMAT = "MMM yyyy";

    private Calendar currentDate;

    private CalendarGridAdapter gridAdapter;

    private ImageView buttonPrevMonth, buttonNextMonth;
    private TextView tvCurrentMonth;
    private GridView calendarGrid;


    public CalendarView(Context context) {
        super(context);
        initGui(context);
    }


    public CalendarView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initGui(context);
    }

    public CalendarView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initGui(context);
    }

    public void setGridAdapter(CalendarGridAdapter adapter) {
        this.gridAdapter = adapter;
        calendarGrid.setAdapter(gridAdapter);
        refresh();
    }

    public void setDayOnClickListener(AdapterView.OnItemClickListener listener) {
        calendarGrid.setOnItemClickListener(listener);
    }

    public void setNextMonthOnClickListener(OnClickListener listener) {
        buttonNextMonth.setOnClickListener(listener);
    }



    public void setPrevMontOnClickListener(OnClickListener listener) {
        buttonPrevMonth.setOnClickListener(listener);

    }


    private void initGui(Context context) {
        LayoutInflater inflater = (LayoutInflater)
                context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        inflater.inflate(R.layout.calendar, this);

        currentDate = Calendar.getInstance();


        buttonPrevMonth = (ImageView) findViewById(R.id.iv_prev_month);
        buttonNextMonth = (ImageView) findViewById(R.id.iv_next_month);
        tvCurrentMonth = (TextView) findViewById(R.id.tv_current_month);
        calendarGrid = (GridView)findViewById(R.id.gv_calendargrid);
    }



    public void refresh() {
        Calendar calendar = (Calendar) currentDate.clone();

        int selectedMonth = calendar.get(Calendar.MONTH);

         // determine cell for the current month's beginning
        calendar.set(Calendar.DAY_OF_MONTH, 1);
        int monthBeginningCell = calendar.get(Calendar.DAY_OF_WEEK) - 1;

        // move calendar backwards to the beginning of the week
        calendar.add(Calendar.DAY_OF_MONTH, -monthBeginningCell);

        // fill cells
        gridAdapter.clear();
        int dayCounter = 0;
        while(dayCounter < MAX_DAY_COUNT) {
            gridAdapter.insert(calendar.getTime(), dayCounter);
            calendar.add(Calendar.DAY_OF_MONTH, 1);
            dayCounter++;
        }
        gridAdapter.setMonth(selectedMonth);
        gridAdapter.notifyDataSetChanged();

        // update title
        tvCurrentMonth.setText(getMonth(currentDate.get(Calendar.MONTH)) + " " + currentDate.get(Calendar.YEAR));

    }

    private String getMonth(int month) {

        switch(month) {
            case 0:
                return getResources().getString(R.string.january);
            case 1:
                return getResources().getString(R.string.february);
            case 2:
                return getResources().getString(R.string.march);
            case 3:
                return getResources().getString(R.string.april);
            case 4:
                return getResources().getString(R.string.mai);
            case 5:
                return getResources().getString(R.string.june);
            case 6:
                return getResources().getString(R.string.july);
            case 7:
                return getResources().getString(R.string.august);
            case 8:
                return getResources().getString(R.string.september);
            case 9:
                return getResources().getString(R.string.october);
            case 10:
                return getResources().getString(R.string.november);
            case 11:
                return getResources().getString(R.string.december);
            default:
                return "UNKNOWN MONTH";
        }
    }

    public void incMonth(int i) {
        currentDate.add(Calendar.MONTH, i);
    }
}

File: app/src/main/java/org/secuso/privacyfriendlytodolist/view/calendar/CalendarView.kt
/*
Privacy Friendly To-Do List
Copyright (C) 2016-2025  Dominik Puellen

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package org.secuso.privacyfriendlytodolist.view.calendar

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.AdapterView.OnItemClickListener
import android.widget.GridView
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import org.secuso.privacyfriendlytodolist.R
import java.util.Calendar

class CalendarView : LinearLayout {
    private var currentDate: Calendar
    private var buttonPrevMonth: ImageView
    private var buttonNextMonth: ImageView
    private var tvCurrentMonth: TextView
    private var daysOfWeek: Array<TextView>
    private var calendarGrid: GridView
    private var monthNames: Array<String>
    private var daysOfWeekNames: Array<String>
    private var gridAdapter: CalendarGridAdapter? = null
    private var firstDayOfWeek = Calendar.SUNDAY

    constructor(context: Context) :
            super(context)

    constructor(context: Context, attrs: AttributeSet?) :
            super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
            super(context, attrs, defStyleAttr)

    init {
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        inflater.inflate(R.layout.calendar, this)
        currentDate = Calendar.getInstance()
        buttonPrevMonth = findViewById(R.id.iv_prev_month)
        buttonNextMonth = findViewById(R.id.iv_next_month)
        tvCurrentMonth = findViewById(R.id.tv_current_month)
        daysOfWeek = arrayOf(
            findViewById(R.id.tv_1_day_of_week),
            findViewById(R.id.tv_2_day_of_week),
            findViewById(R.id.tv_3_day_of_week),
            findViewById(R.id.tv_4_day_of_week),
            findViewById(R.id.tv_5_day_of_week),
            findViewById(R.id.tv_6_day_of_week),
            findViewById(R.id.tv_7_day_of_week))
        calendarGrid = findViewById(R.id.gv_calendar_grid)

        monthNames = arrayOf(
            resources.getString(R.string.january),
            resources.getString(R.string.february),
            resources.getString(R.string.march),
            resources.getString(R.string.april),
            resources.getString(R.string.may),
            resources.getString(R.string.june),
            resources.getString(R.string.july),
            resources.getString(R.string.august),
            resources.getString(R.string.september),
            resources.getString(R.string.october),
            resources.getString(R.string.november),
            resources.getString(R.string.december))
        daysOfWeekNames = arrayOf(
            resources.getString(R.string.sunday_abbr),
            resources.getString(R.string.monday_abbr),
            resources.getString(R.string.tuesday_abbr),
            resources.getString(R.string.wednesday_abbr),
            resources.getString(R.string.thursday_abbr),
            resources.getString(R.string.friday_abbr),
            resources.getString(R.string.saturday_abbr))

        setFirstDayOfWeek(firstDayOfWeek)
    }

    fun setGridAdapter(adapter: CalendarGridAdapter?) {
        gridAdapter = adapter
        calendarGrid.setAdapter(gridAdapter)
    }

    /**
     * Sets the first day of a week for this calendar.
     *
     * @param day A weekday constant from [Calendar] class, e.g. [Calendar.MONDAY].
     */
    fun setFirstDayOfWeek(day: Int) {
        if (day >= Calendar.SUNDAY && day <= Calendar.SATURDAY) {
            firstDayOfWeek = day
            var dayNameIndex = firstDayOfWeek - 1
            for (dayViewIndex in 0..6) {
                daysOfWeek[dayViewIndex].text = daysOfWeekNames[dayNameIndex]
                dayNameIndex = (dayNameIndex + 1) % 7
            }
        }
    }

    fun setDayOnClickListener(listener: OnItemClickListener?) {
        calendarGrid.onItemClickListener = listener
    }

    fun setNextMonthOnClickListener(listener: OnClickListener?) {
        buttonNextMonth.setOnClickListener(listener)
    }

    fun setPrevMontOnClickListener(listener: OnClickListener?) {
        buttonPrevMonth.setOnClickListener(listener)
    }

    fun refresh() {
        val calendar = currentDate.clone() as Calendar
        val selectedYear = calendar[Calendar.YEAR]
        val selectedMonth = calendar[Calendar.MONTH]

        // determine cell for the current month's beginning
        calendar[Calendar.DAY_OF_MONTH] = 1
        val monthBeginningCell = (7 + calendar[Calendar.DAY_OF_WEEK] - firstDayOfWeek) % 7

        // move calendar backwards to the beginning of the week
        calendar.add(Calendar.DAY_OF_MONTH, -monthBeginningCell)

        // update title
        val text = getMonthName(currentDate[Calendar.MONTH]) + " " + currentDate[Calendar.YEAR]
        tvCurrentMonth.text = text

        // fill cells
        val adapter = gridAdapter ?: return
        adapter.clear()
        for (dayIndex in 0..<MAX_DAY_COUNT) {
            adapter.insert(calendar.time, dayIndex)
            calendar.add(Calendar.DAY_OF_MONTH, 1)
        }
        adapter.setSelectedDate(selectedYear, selectedMonth)
        adapter.notifyDataSetChanged()
    }

    private fun getMonthName(month: Int): String {
        return if (month in monthNames.indices) monthNames[month] else "UNKNOWN MONTH $month"
    }

    fun incMonth(i: Int) {
        currentDate.add(Calendar.MONTH, i)
    }

    companion object {
        private const val MAX_DAY_COUNT = 42
    }
}
